File: 14.longest_common_prefix/prefix.go
// 14. Longest common prefix
// Topics: 'Array', 'String', 'Trie'

// Write a function to find the longest common prefix string amongst an array of strings.

// If there is no common prefix, return an empty string "".

// Example 1:

// Input: strs = ["flower","flow","flight"]
// Output: "fl"

// Example 2:

// Input: strs = ["dog","racecar","car"]
// Output: ""
// Explanation: There is no common prefix among the input strings.

// Constraints:

//     1 <= strs.length <= 200
//     0 <= strs[i].length <= 200
//     strs[i] consists of only lowercase English letters if it is non-empty.

package longestcommonprefix

func longestCommonPrefix(strs []string) string {
	if len(strs) == 1 {
		return strs[0]
	}
	prefix := len(strs[0])
	for _, str := range strs[1:] {
		var tempPref int
		for i, s := range strs[0][:prefix] {
			if i >= len(str) {
				break
			}
			if byte(s) == str[i] {
				tempPref++
			} else {
				break
			}
		}
		if tempPref == 0 {
			return ""
		}
		if prefix > tempPref {
			prefix = tempPref
		}
	}

	return strs[0][:prefix]
}

File: 2185.counting_words_with_a_given_prefix/prefix.py
# 2185. Counting words with a given prefix
# Topics: 'Array', 'String', 'String Matching'

# You are given an array of strings words and a string pref.

# Return the number of strings in words that contain pref as a prefix.

# A prefix of a string s is any leading contiguous substring of s.

 

# Example 1:

# Input: words = ["pay","attention","practice","attend"], pref = "at"
# Output: 2
# Explanation: The 2 strings that contain "at" as a prefix are: "attention" and "attend".

# Example 2:

# Input: words = ["leetcode","win","loops","success"], pref = "code"
# Output: 0
# Explanation: There are no strings that contain "code" as a prefix.

 

# Constraints:

#     1 <= words.length <= 100
#     1 <= words[i].length, pref.length <= 100
#     words[i] and pref consist of lowercase English letters.

from typing import List

class Node:
    def __init__(self):
        self.children = {}
        self.times = 1

class Trie:
    def __init__(self):
        self.root = Node()
    
    def insert(self, word: str, length: int):
        cur = self.root
        for i in range (length):
            if word[i] not in cur.children:
                cur.children[word[i]] = Node()
            else:
                cur.children[word[i]].times += 1
            cur = cur.children[word[i]]
    
    def prefix(self, pref: str) -> int:
        cur = self.root
        for char in pref:
            if char not in cur.children:
                return 0
            cur = cur.children[char]
        return cur.times

class Solution:
    def prefixCount(self, words: List[str], pref: str) -> int:
        trie = Trie()
        for word in words:
            if len(word) >= len(pref):
                trie.insert(word, len(pref))
        return trie.prefix(pref)
